streaming/dataset.py

```diff
@@ -450,7 +449,16 @@ def __init__(self,
         # instantiate a World inside the StreamingDataset destructor.
         self._rank_world = world = World()
 
-        # Download each stream's index, load their shards, and map streams <-> shards.
+        # Download each Stream's index in parallel from local rank 0.
+        #
+        # Parallelism is important because there could be a very large number of Streams, and we
+        # expect equal performance as them having been concatenated into one Stream.
+        if world.is_local_leader:
+            with Pool() as pool:
+                pool.imap_unordered(lambda stream: stream.download_index(), self.streams)
```

Comment:
Do we need to call the `wait()` method from `pool,` which blocks until all tasks issued asynchronously are completed?

Reply:
Ah yea: https://docs.python.org/3.7/library/multiprocessing.html#multiprocessing.pool.Pool.join

Fixed in this commit: https://github.com/mosaicml/streaming/pull/585/commits/25c8f152dc2cd61e21edc97b8c517441aa5d94e6

Where I also add: `procs = SD._get_index_download_procs(procs, procs / cpu, max procs)`

Note: the Parquet PR is going to add indexing to SD init -- I suspect the above control on parallelism will get merged into that -- I wouldn't take this code as the last word